src/shared/util.js

```diff
@@ -710,19 +710,27 @@ class Util {
   // The SVD can be represented with formula A = USV. We are interested in the
   // matrix S here because it represents the scale values.
   static singularValueDecompose2dScale(m) {
-    const transpose = [m[0], m[2], m[1], m[3]];
+    if (Math.abs(m[1]) <= Number.EPSILON && Math.abs(m[2]) <= Number.EPSILON) {
+      // Diagonal matrix
+      return [Math.abs(m[0]) || 1, Math.abs(m[3]) || 1];
+    }
+
+    if (Math.abs(m[0]) <= Number.EPSILON && Math.abs(m[3]) <= Number.EPSILON) {
+      // Anti-diagonal matrix: scale composed with rotation of +/- 90°
+      return [Math.abs(m[2]) || 1, Math.abs(m[1]) || 1];
+    }
 
     // Multiply matrix m with its transpose.
-    const a = m[0] * transpose[0] + m[1] * transpose[2];
-    const b = m[0] * transpose[1] + m[1] * transpose[3];
-    const c = m[2] * transpose[0] + m[3] * transpose[2];
-    const d = m[2] * transpose[1] + m[3] * transpose[3];
+    const a = m[0] * m[0] + m[1] * m[1];
+    const b = m[0] * m[2] + m[1] * m[3];
+    const c = m[2] * m[0] + m[3] * m[1];
+    const d = m[2] * m[2] + m[3] * m[3];
 
     // Solve the second degree polynomial to get roots.
-    const first = (a + d) / 2;
-    const second = Math.sqrt((a + d) * (a + d) - 4 * (a * d - c * b)) / 2;
-    const sx = first + second || 1;
-    const sy = first - second || 1;
+    const sum_a_d = a + d;
+    const second = Math.sqrt(sum_a_d * sum_a_d - 4 * (a * d - c * b));
```

Comment:
```suggestion
    const second = Math.sqrt(sum_a_d ** 2 - 4 * (a * d - c * b));
```

Also, the variable name `second` looks slightly out-of-place here now.